1-Draft/RFCNNNN-ML-Based-Command-Line-Completion.md

```diff
@@ -0,0 +1,174 @@
+
+### Goals
+
+* Support for both local and cloud based implementations of any component
+* Integrates with existing PowerShell command completion
+* Provides for privacy protection
+* Supports feature opt-in
+
+## Specification 
+
+### Modules
+#### Prediction Module  
+* Collects command line text used in PowerShell for model training
+* Returns prediction from submitted command line fragment
+* Creates suggestion success statistics based on user choice
+
+#### General Data Protection Regulation (GDPR) support Module
+These regulations apply to any personal data that is collected, and allows the user to:
+* View
+* Edit
+* Delete
+* Restrict
+* Export
+
+The mechanism for feeding the model with data and retrieving results will closely follow the existing command completion implementation.  The Prediction Module will implement APIs that will:  
+* Accept command line fragments with which to generate predictions
+* Provide prediction information including confidence level
+* Collect complete command lines processed by PowerShell (GDPR applied)
+* Generate success statistics based on user choices
+
+The user experience will be dependent on the PowerShell host and is outside the scope of this RFC. But hosts will need to be updated to take advantage of the new opt-in prediction feature. For example PowerShell shell PSReadline should integrate it into its tab completion and/or Ctrl+Space suggestion list display. VSCode should likewise integrate it into its tab completion and intellisense.  
+
+Because of potential lag concerns (especially when the machine learning model is implemented in the cloud), querying prediction results will be an asynchronous operation.  
+
+### Interfaces
+
+```
+     public sealed class CommandLinePredictionResult
+    {
+        /// <summary>
+        /// Command line text suggested by the ML model
+        /// </summary>
+        public string SuggestionText { get; }
+
+        /// <summary>
+        /// Value indicating confidence level of prediction (0-100)
+        /// </summary>
+        public int SuggestionConfidence { get; }
+
+        /// <summary>
+        /// Null instance when no prediction is made
+        /// </summary>
+        public static readonly CommandLinePredictionResult NullInstance = new CommandLinePredictionResult();
+    }
+
+    public sealed class CommandLinePredictionStatistics
+    {
+        // TBD Prediction success statistics
+    }
+
+    interface CommandLinePredictor
+    {
+        /// <summary>
+        /// Performs machine learning prediction on provided command line fragment
+        /// </summary>
+        /// <param name="input">Command line fragment</param>
+        /// <param name="cursorIndex">Cursor index</param>
+        /// <param name="OnPredictionCallback">Call back delegate to receive prediction results</param>
+        void PredictFromInput(string input, int cursorIndex, Action<Collection<CommandLinePredictionResult>> OnPredictionCallback);

```

Comment:
I suggest using the asynchronous programming with async and await in C#. The signature would look like:
```c#
async Task<Collection<CommandLinePredictionResult>> GetPredictFromInput(string input, int cursorIndex)
```

Reply:
Sounds good.  Will change.